### PR TITLE
blockdev_stream_cor_base: stream with cor filter as base

### PR DESCRIPTION
**Pull request by aliang123**
Block stream with base: copy-on-read node

Signed-off-by: Aihua Liang <aliang@redhat.com>
id:2017292

### CODE REVIEW

**Comment by ngu-niny** [comment]
(1/4) Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.6.0.ppc64le.io-github-autotest-qemu.blockdev_stream_cor_base: PASS (49.19 s)

 (2/4) Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.ppc64le.io-github-autotest-qemu.blockdev_stream_cor_base: PASS (53.84 s)

Acked-by: Nini Gu <ngu@redhat.com>

File: qemu/tests/cfg/blockdev_stream_cor_base.cfg [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+    only Linux
+    start_vm = yes
+    kill_vm = yes
+    qemu_force_use_drive_expression = no
+    type = blockdev_stream_cor_base
+    virt_test_type = qemu
+    base_tag = image1
+    node = drive_${base_tag}
+    base_node = drive_${base_tag}
+    image_copy_on_read = True
```
Comment (zhencliu) [inline qemu/tests/cfg/blockdev_stream_cor_base.cfg:11]: still, yes or on or true

File: qemu/tests/blockdev_stream_cor_base.py [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+        self.blockdev_stream()
+
+
+def run(test, params, env):
+    """
+    Basic block stream with cor filter node as base
+
+    test steps:
+        1. boot VM with copy-on-read filter
+        2. create snapshot chains base->cor filter->sn1->sn2->sn3->sn4
```
Comment (zhencliu) [inline qemu/tests/blockdev_stream_cor_base.py:96]: base(cor filter is the base)->sn1->sn2->sn3->sn4

File: qemu/tests/blockdev_stream_cor_base.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+
+class BlockdevStreamCORBase(BlockDevStreamTest):
+    """Do block-stream with copy-on-read filter as base """
+
+    def __init__(self, test, params, env):
+        super(BlockdevStreamCORBase, self).__init__(test, params, env)
+        self.snapshot_chain = params["snapshot_chain"].split()
+
+    def prepare_snapshot_file(self):
+        self.trash = []
```
Comment (zhencliu) [inline qemu/tests/blockdev_stream_cor_base.py:19]: add snapshot images into trash, but never remove them

File: qemu/tests/blockdev_stream_cor_base.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+        cmd = "blockdev-snapshot"
+        arguments = self.params.copy_from_keys(options)
+        for snapshot_tag in self.snapshot_chain:
+            arguments["overlay"] = "drive_%s" % snapshot_tag
+            self.main_vm.monitor.cmd(cmd, dict(arguments))
+            arguments["node"] = "drive_%s" % snapshot_tag
+            if snapshot_tag != self.snapshot_chain[-1]:
+                self.backing_chain.append(snapshot_tag)
+        self._top_device = "drive_%s" % self.snapshot_chain[-1]
+        self.check_backing_chain()
```
Comment (zhencliu) [inline qemu/tests/blockdev_stream_cor_base.py:75]: Do we need to do this check? after stream, we can check if sn4's backing is the system image, here, this checking should be covered in the snapshot related cases

File: qemu/tests/blockdev_stream_cor_base.py [hunk 5]
```diff
@@ -0,0 +1,106 @@
...
+
+    def blockdev_stream(self):
+        backup_utils.blockdev_stream(self.main_vm, self._top_device,
+                                     **self._stream_options)
+        time.sleep(0.5)
+        index = self.backing_chain.index(self.params["base_tag"])
+        self.backing_chain = self.backing_chain[:index]
```
Comment (zhencliu) [inline qemu/tests/blockdev_stream_cor_base.py:82]: is index = 0? then self.backing_chain = []?

File: qemu/tests/cfg/blockdev_stream_cor_base.cfg (hunk 1 above)
Comment (aliang123) [reply qemu/tests/cfg/blockdev_stream_cor_base.cfg]: Have updated, thanks zhenchao.

File: qemu/tests/blockdev_stream_cor_base.py (hunk 2 above)
Comment (aliang123) [reply qemu/tests/blockdev_stream_cor_base.py]: Updated to base(cor filter attached)->sn1->sn2->sn3->sn4

File: qemu/tests/blockdev_stream_cor_base.py (hunk 3 above)
Comment (aliang123) [reply qemu/tests/blockdev_stream_cor_base.py:19]: Have updated, remove image in post_test. Thanks zhenchao.

File: qemu/tests/blockdev_stream_cor_base.py (hunk 5 above)
Comment (aliang123) [reply qemu/tests/blockdev_stream_cor_base.py]: Have updated, just keep elments with index <=0 in self.backing_chain.

**Comment by aliang123** [comment]
(1/1) Host_RHEL.m9.u0.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.9.0.0.x86_64.io-github-autotest-qemu.blockdev_stream_cor_base.q35: PASS (41.72 s)

File: qemu/tests/blockdev_stream_cor_base.py (hunk 4 above)
Comment (aliang123) [reply qemu/tests/blockdev_stream_cor_base.py:75]: Checking the backing file info to make sure the cor filter attached to base node after snapshot and stream ops.

**Comment by ngu-niny** [comment]
1-Host_RHEL.m8.u6.product_rhel.qcow2.virtio_scsi.up.virtio_net.Guest.RHEL.8.6.0.ppc64le.io-github-autotest-qemu.blockdev_stream_cor_base Finshed 2021-11-07 20:40:59 2021-11-07 20:42:01 PASS 62.50618

**Pull request merged**